packages/framework/attributor/src/attributor.ts

```diff
@@ -0,0 +1,109 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { assert } from "@fluidframework/common-utils";
+import { IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { ISequencedDocumentMessage, IUser } from "@fluidframework/protocol-definitions";
+import { UsageError } from "@fluidframework/container-utils";
+
+/**
+ * Attribution information associated with a change.
+ */
+export interface AttributionInfo {
+	/**
+	 * The user that performed the change.
+	 */
+	user: IUser;
+	/**
+	 * When the change happened.
+	 */
+	timestamp: number;
+}
+
+/**
+ * Provides lookup between attribution keys and their associated attribution information.
+ */
+export interface IAttributor {
+	/**
+	 * Retrieves attribution information associated with a particular key.
+	 * @param key - Attribution key to look up.
+	 * @throws If no attribution information is recorded for that key.
+	 */
+	getAttributionInfo(key: number): AttributionInfo;
+
+	/**
+	 * @param key - Attribution key to look up.
+	 * @returns the attribution information associated with the provided key, or undefined if no information exists.
+	 */
+	tryGetAttributionInfo(key: number): AttributionInfo | undefined;
+
+	/**
+	 * @returns an iterable of (attribution key, attribution info) pairs for each stored key.
+	 */
+	entries(): IterableIterator<[number, AttributionInfo]>;
+
+	// TODO:
+	// - GC
+}
+
+/**
+ * {@inheritdoc IAttributor}
+ */
+export class Attributor implements IAttributor {
+	protected readonly keyToInfo: Map<number, AttributionInfo>;
+
+	/**
+	 * @param initialEntries - Any entries which should be populated on instantiation.
+	 */
+	constructor(
+		initialEntries?: Iterable<[number, AttributionInfo]>,
+	) {
+		this.keyToInfo = new Map(initialEntries ?? []);
+	}
+
+	/**
+	 * {@inheritdoc IAttributor.getAttributionInfo}
+	 */
+	public getAttributionInfo(key: number): AttributionInfo {
+		const result = this.tryGetAttributionInfo(key);
+		if (!result) {
+			throw new UsageError(`Requested attribution information for unstored key: ${key}.`);
+		}
+		return result;
+	}
+
+	/**
+	 * {@inheritdoc IAttributor.tryGetAttributionInfo}
+	 */
+	public tryGetAttributionInfo(key: number): AttributionInfo | undefined {
+		return this.keyToInfo.get(key);
+	}
+
+	/**
+	 * {@inheritdoc IAttributor.entries}
+	 */
+	public entries(): IterableIterator<[number, AttributionInfo]> {
+		return this.keyToInfo.entries();
+	}
+}
+
+/**
+ * Attributor which listens to an op stream and records entries for each op.
+ * Sequence numbers are used as attribution keys.
+ */
+export class OpStreamAttributor extends Attributor implements IAttributor {
+	constructor(
+		runtime: IFluidDataStoreRuntime,
```

Comment:
I'm working on changes needed to actually wire this up now and think this probably isn't right--leaning more towards ContainerRuntime level, but even with that it's probably better to only demand what we need here (an audience and something we can listen to for ops)

Reply:
I have plans to change this to only take in what it needs (audience object and delta manager), which makes mocking easier. I guess that didn't make it into this PR when I split it off of the large set of changes. I'd prefer to leave it this way for now if there aren't strong objections (note the package is still private, can iterate fast)